mkosi/backend.py

```diff
@@ -955,3 +957,22 @@ def complete_step(cls, text: str, text2: Optional[str] = None) -> Iterator[List[
 
         if text2 is not None:
             cls.print_step(text2.format(*args))
+
+
+def chown_to_running_user(paths: Union[PathString, [PathString]]):
+    uid = int(os.getenv("SUDO_UID") or os.getenv("PKEXEC_UID") or '0')
+    user = pwd.getpwuid(uid).pw_name
+    gid = pwd.getpwuid(uid).pw_gid
+
+    # Root already owns generated files/directories, skipping.
+    if uid == 0:
+        return
+
+    if type(paths) is not list: paths = [ paths ]
+
+    for path in paths:
+        with MkosiPrinter.complete_step(
+            f"Changing ownership of output file {path} to user {user}…",
+            f"Changed ownership of {path}",
+        ):
+            os.chown(path, uid, gid)
```

Comment:
i don't grok the relationship to --no-chown?

it appears to me that there should be only one option: either we do the chowning to the sudo/pkexec user or we do not. other than that we should probably always use the new `rootidmap` setting if nspawn supports it for everything we bind mount from the build tree.

to summarize:
1. if chowning is on, and nspawn supports rootidmap: always chown dirs we create, and use rootidmap when passing them to the container
2. otherwise let dirs we created owned by root, and do not use rootidmap
3. if chowning is on, chown output files

That would mean one option less, and we should retain compat with status quo ante mkosi trees.

Reply:
The relation with `--no-chown` is the following:
- If `--no-chown` is used: `--map-permissions` can't be used, this is checked in `load_args()`).
- If `--no-chown` is **not** used and `--map-permission` is **not** used: no change from previous behaviour.
- If `--no-chown` is **not** used and `--map-permissions` is used: created directories are `chown`ed to `SUDO_UID` or `PKEXEC_UID`, and `rootidmap` is used. This way, this would preserve the existing behaviour.

Following your and Daan's comments:
- Remove `--map-permissions` and used `rootidmap` by default if nspawn supports it. This would be OK whatever the owner of the directory is, the worse thing is it will remap `root` to `root` if the owner of the `--bind` mounted directory is `root`.
- By default, `chown` all newly created build directories, except if `--no-chown` is used.

Reply:
Updated the PR's details with the latest changes and explanations.